integration_tests/src/main/python/asserts.py

```diff
@@ -678,34 +678,79 @@ def do_it_all(spark):
     :param error_message: a string such as the one produce by java.lang.Exception.toString
-    :return: Assertion failure if no exception matching error_message has occurred.
+    Assertion failure if:
+        - no exception matching error_message has occurred
+        - the error isn't a string or a regex
     """
-    with pytest.raises(Exception) as excinfo:
-        func()
-    actual_error = excinfo.exconly()
     if isinstance(error_message, re.Pattern):
         assert error_message.search(actual_error), f"Expected error '{error_message}' to match '{actual_error}'"
     elif isinstance(error_message, str):
         assert error_message in actual_error, f"Expected error '{error_message}' did not appear in '{actual_error}'"
     else:
         assert False, f"expected str or Pattern found {type(error_message)} {error_message}"
 
-def assert_gpu_and_cpu_error(df_fun, conf, error_message):
+
+def assert_spark_exception(func, error_message, expect_exception=None):
+    """
+    Assert that a specific Java exception is thrown
+    :param func: a function to be verified
+    :param error_message: a string such as the one produce by java.lang.Exception.toString
+    :param expect_exception:
+            - if none, we check the error if and only if it was raised
+            - if true, we will assert False if func() doesn't raise an exception,
+            - if false, we fail if we did not raise an exception
+
+    :return: (true, None) if the code failed with an exception and the text matches `error_message`.
+             (false, result) if `expect_exception` is false and the function doesn't throw
+             Assertion failure if:
+             - no exception matching error_message has occurred, or
+             - the code is expected to throw and doesn't, or
+             - the code throws and is not expected to throw.
+    """
+    thrown_ex = None
+    result = None
+    try:
+        result = func()
+    except Exception as ex:
+        thrown_ex = str(ex)
+
+    if thrown_ex is not None:
+        if expect_exception is not False:
+            check_exception(thrown_ex, error_message)
+            return (True, None)
+        else:
+            assert False, f"no exception was expected, but the test failed with {thrown_ex}"
+
+    elif expect_exception:
+        # we expected an exception, but func() didn't throw
+        assert False, f"expected exception with error message {error_message}, but the test did not throw."
+    return (False, result)
+
+
+def assert_gpu_and_cpu_error(df_fun, conf, error_message, only_if_cpu_fails=False):
     """
     Assert that GPU and CPU execution results in a specific Java exception thrown
     :param df_fun: a function to be verified
     :param conf: Spark config
     :param error_message: a string such as the one produce by java.lang.Exception.toString
+    :param only_if_cpu_fails: if true, we only expect the GPU job to fail if the CPU job fails,
```

Comment:
nit: for readability I would rather have this boolean not be here, but instead have a new method that would be something more like. `assert_gpu_and_cpu_same_data_or_error`, but I don't really like that name too much either.